### DISCUSSION TITLE
Only support convex polygon?

### DISCUSSION THREAD

**Discussion by wxfred** [kaisalmen/WWOBJLoader#73] [Ideas]
Is there any plan to support concave polygon without hole?

**Comment by kaisalmen** [comment]
Hi @wxfred could you elaborate more what you mean? 
Do you have an example `.obj` file showing your problem? Thank you

**Comment by wxfred** [comment]
Search **earcut** on npm, you will know what i mean.

**Reply by kaisalmen** [reply]
That is a triangulation library. This is a loader for the OBJ file format which describes triangles/mesh data, etc.
Are you talking about a post-process? What is your idea and please describe it?  Thanks

**Reply by wxfred** [reply]
The obj file exported from Blender may have some concave polygon faces, unfortunately, your obj loader will parse it with some errors.

![concave_origin](https://user-images.githubusercontent.com/6032500/233777749-2852ee05-74c5-460f-b0e3-c57923c267ac.jpg)
![concave_error](https://user-images.githubusercontent.com/6032500/233777755-49491320-bb1a-48b3-a587-d37c05286a54.jpg)

the correct triangulation should be like this

![concave_correct](https://user-images.githubusercontent.com/6032500/233777799-c5968d7d-be90-432e-9e90-515b174119e4.jpg)

So, do the triangulation by using the earcut will be nice.

**Reply by kaisalmen** [reply]
Ok, thank you. That's something I can work with. I wasn't aware there is specific problem. So, either the OBJ contains errors or the OBJLoader2 does it wrong (more likely).

**Comment by kaisalmen** [comment]
@wxfred https://github.com/kaisalmen/WWOBJLoader/pull/74 should fix this problem.
I have published a preview version based on the PR:
https://www.npmjs.com/package/wwobjloader2/v/6.0.1-next.0

Please, let me know if this works as expected. Thank you for reporting the issue.

**Reply by kaisalmen** [reply]
@wxfred did you have time to test the pre-release?

**Reply by wxfred** [reply]
I tested, still wrong.
"three": "^0.152.2"
"wwobjloader2": "^6.0.1-next.0"

Github is not support uploading .obj format, you can download my testing file below and change the extname.
[tritest.txt](https://github.com/kaisalmen/WWOBJLoader/files/11471572/tritest.txt)

my testing code
```
// utils.js
import axios from 'axios'
import { OBJExporter } from 'three/examples/jsm/exporters/OBJExporter'
import { OBJLoader2 } from 'wwobjloader2'

function blob2TxtPromise(blob) {
  return new Promise((resolve) => {
    const reader = new FileReader()
    reader.readAsText(blob, 'utf8')
    reader.onload = () => {
      resolve(reader.result)
    }
  })
}

async function importObj(url) {
  const { data: blob } = await axios.get(url, { responseType: 'blob' })
  const txt = await blob2TxtPromise(blob)
  const objLoader2 = new OBJLoader2()
  return objLoader2.parse(txt)
}

function exportObj(object3D) {
  const exporter = new OBJExporter()
  return exporter.parse(object3D)
}

export default {
  importObj,
  exportObj,
}
```

the original model in blender
![image](https://github.com/kaisalmen/WWOBJLoader/assets/6032500/0ef69260-a464-47db-b21c-a0ef6fef4f32)

and import it by your code then export to blender
![image](https://github.com/kaisalmen/WWOBJLoader/assets/6032500/3c111985-e9be-44ee-9d03-adfed65a0145)

**Reply by wxfred** [reply]
The one of the correct triangulation is like this
![image](https://github.com/kaisalmen/WWOBJLoader/assets/6032500/bc009e4a-9bdd-4a3f-837f-ee12e74f1b02)

**Reply by wxfred** [reply]
I uploaded a package named three-obj-earcut, which is doing triangulation on obj txt level (means importing progress needs parsing the obj document twice, from obj txt to triangulated obj txt, then to object3D).
```
// utils.js
import axios from 'axios'
import { OBJExporter } from 'three/examples/jsm/exporters/OBJExporter'
import triangulate from 'three-obj-earcut' // change
import { OBJLoader2 } from 'wwobjloader2'

function blob2TxtPromise(blob) {
  return new Promise((resolve) => {
    const reader = new FileReader()
    reader.readAsText(blob, 'utf8')
    reader.onload = () => {
      resolve(reader.result)
    }
  })
}

async function importObj(url) {
  const { data: blob } = await axios.get(url, { responseType: 'blob' })
  const txt = triangulate(await blob2TxtPromise(blob)) // change
  const objLoader2 = new OBJLoader2()
  return objLoader2.parse(txt)
}

function exportObj(object3D) {
  const exporter = new OBJExporter()
  return exporter.parse(object3D)
}

export default {
  importObj,
  exportObj,
}
```

**Reply by kaisalmen** [reply]
@wxfred what I tried only works in very specific cases and cannot generally work with concave polygons. I checked to OBJ spec again and it only allows to define multiple faces (`f`) as already supported by `OBJLoader2`. The main three.js `OBJLoader` works the same way. I guess that Blender on re-import of the OBJ applies the same logic for re-creating the surfaces it used for writing it out and therefore it works. If you import the file somewhere else you will very likely end up with the same problem (see here: https://github.com/godotengine/godot/issues/27447)

The only stable way is to set "Triangulated Mesh" when exporting the file with the Blender OBJ exporter. Then `OBJLoader2` loads the mesh correctly:
![image](https://github.com/kaisalmen/WWOBJLoader/assets/21079138/ef063c28-6949-40c2-b6fc-efec96aa8dfa)

I will not merge and close the proposed PR #74 .

**Reply by wxfred** [reply]
The OBJLoader of three.js shouldn't be treated as a standard, it even recalculate the normal coordinates instead of using the vertex normals in the obj file! If you have learned OpenGL or WebGL (which is the fundamental of three.js), you will know the difference between vertex normal and face normal, OBJLoader recalculates face normal and overwrites the result to vertices of that face. In short, the orignal OBJLoader is a partial-featured obj parser.
Another example, the vtk model parser in three.js is also a partial-featured vtk parser. Vtk format supports many types of structure, but they only implement the simplest one.

I searched npm, found your package, thought you may have the ambition to implement a full-featured obj parser, maybe I was wrong.
It's up to you.

By the way, my package is not full-featured, as I mentioned in README, I assume that there are vertex normal information (like vn 0 0 1) in the obj file, but there is no requirement that the obj file must have normal information. A full-featured parser must consider everything.

**Reply by kaisalmen** [reply]
> I searched npm, found your package, thought you may have the ambition to implement a full-featured obj parser, maybe I was wrong.
It's up to you.

Like many other OBJ parser this parser is not fully featured (e.g. free-form curves), but with regard to standard polygonal mesh related things it does a fairly good job (this is a spare time project).

The important point here is that the faces description in the OBJ specification (http://paulbourke.net/dataformats/obj/) **only** allows to specify triangle fans. Do you have any other information I am not aware of? There is no way to distinguish between concave and convex polygons in the OBJ file that's why you have to use triangulation before exporting as suggested above OR
we could make this library depend on earcut and add a parser option to use it.

Feel free to offer support and open a PR. Thanks

**Reply by wxfred** [reply]
I didn't see any **fan** description in [http://paulbourke.net/dataformats/obj/](http://paulbourke.net/dataformats/obj/), pls paste them.
If there is a fan statement in obj format standard, I will admit I was wrong.
Whatever, Blender supports importing concave face obj without triangulating, I think concave parse supporting is really helpful.

**Reply by kaisalmen** [reply]
You build faces from referencing other things, like vertices, uvs and normals. This the simplest case with only referencing vertices:
```
f 1 2 3 4 5 6
```
All cases here: https://github.com/kaisalmen/WWOBJLoader/blob/main/packages/objloader2/src/OBJLoader2Parser.ts#L396-L435).

You build the following faces/triangles **1,2,3; 1,3,4; 1,4,5; 1.5.6** from that. The spec doesn't give you any further hints regarding ordering or structure of the data. So from just looking at what you see, you cannot derive a higher ordering/structure or whatsoever. 

Blender seems to interpret the data the same way when exporting and importing again, but that doesn't help you with other software if the spec isn't specific about it (this is independent of OBJLoader2).

OBJ format is super old. Are you using it for a specific reason? Doesn't it make more sense to move to GLTF these days if your starting point is Blender?

**Reply by wxfred** [reply]
I don't think **f 1 2 3 4 5 6** means you should convert this face to triangle fans. In OpenGL, you can use GL_TRIANGLE_STRIP (same as THREE.TriangleStripDrawMode in three.js) to present some kinds of concave face.
![image](https://github.com/kaisalmen/WWOBJLoader/assets/6032500/57aba237-e235-4782-9839-71e9bd2c8438)
You assume **f 1 2 3 4 5 6** in obj file should be parsed to GL_TRIANGLE_FAN, but in my opinion the **f 1 2 3 4 5 6** in obj file is only describe a face surrounded by p1 p2 p3 p4 p5 p6 with a clockwise or counter-clockwise order. This is the point where we disagree with each other.

And, yes, in my recent project, I use obj to do some other things, like physical calculations, I didn't choose GLTF because some CADs is not support this format.

**Reply by kaisalmen** [reply]
@wxfred feel free to propose a PR that adds an option for altering the interpretation (fans or strips) of faces.
Use "Triangulated Mesh" as workaround for now.
